File: app/src/main/java/igrek/songbook/service/songtree/SongTreeItem.java
package igrek.songbook.service.songtree;

import igrek.songbook.domain.songsdb.Song;
import igrek.songbook.domain.songsdb.SongCategory;

public class SongTreeItem {
	
	private Song song;
	private SongCategory category;
	
	private SongTreeItem(Song song, SongCategory category) {
		this.song = song;
		this.category = category;
	}
	
	public static SongTreeItem song(Song song) {
		return new SongTreeItem(song, null);
	}
	
	public static SongTreeItem category(SongCategory category) {
		return new SongTreeItem(null, category);
	}
	
	public String getSimpleName() {
		if (isCategory()) {
			return category.getDisplayName();
		} else {
			return song.getTitle();
		}
	}
	
	public boolean isCategory() {
		return !isSong();
	}
	
	public boolean isSong() {
		return song != null;
	}
	
	public Song getSong() {
		return song;
	}
	
	public SongCategory getCategory() {
		return category;
	}
	
	@Override
	public String toString() {
		if (isCategory()) {
			return "[" + category.getDisplayName() + "]";
		} else {
			return "" + song.getCategory().getDisplayName() + " - " + song.getTitle();
		}
	}
}

File: app/src/main/java/igrek/songbook/songselection/tree/SongTreeItem.kt
package igrek.songbook.songselection.tree

import igrek.songbook.persistence.model.Category
import igrek.songbook.persistence.model.Song

open class SongTreeItem protected constructor(open val song: Song?, val category: Category?) {

    val simpleName: String?
        get() = if (isCategory) {
            category?.displayName
        } else {
            song?.title
        }

    val isCategory: Boolean
        get() = !isSong

    val isSong: Boolean
        get() = song != null

    override fun toString(): String {
        return if (isCategory) {
            "[" + category?.displayName + "]"
        } else {
            "" + song?.categories?.joinToString(", ") { c -> c.displayName!! } + " - " + song?.title
        }
    }

    companion object {

        fun song(song: Song): SongTreeItem {
            return SongTreeItem(song, null)
        }

        fun category(category: Category): SongTreeItem {
            return SongTreeItem(null, category)
        }
    }
}
